File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        Map<String, String> map = new HashMap<>();
        for(String s : phone_book){
            map.put(s, s);
        }
        
        for(String s : phone_book){
            int size = s.length();
            for(int i=1; i<size; i++){
                //System.out.println(s.substring(0,i));
                if(map.get(s.substring(0, i)) != null){
                    answer = false;
                    break;
                }
            }
            if(!answer){
                break;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    
    len_set = set([len(i) for i in phone_book])
    
    check = {}  
    
    for phone_num in phone_book:
        for k in len_set:
            if len(phone_num) <= k:
                continue
            head = phone_num[0:k]
            check[head] = 1
    
    for phone_num in phone_book:
        if check.get(phone_num) != None:
            answer = False
            break
        
    
    return answer